Feature: Support Arbitrary

AlisCode opened the pull request:
Closes #53.

Supports the [`Arbitrary`](https://docs.rs/crate/arbitrary) crate for `NonEmpty`.

Note: I put the target version of Arbitrary to "1", but the latest version is 1.3. Not sure what your policy is on that. 
Ping @greyblake.

greyblake commented on Cargo.toml:
```diff
@@ -9,9 +9,11 @@ repository = "https://github.com/cloudhead/nonempty"
 
 [dependencies]
 serde = { features = ["serde_derive"], optional = true, version = "1" }
+arbitrary = { features = ["derive"], optional = true, version = "1" }
```
I don't think we need `derive` feature

AlisCode replied:
That is specifically used for the test case, to provide a usage example. I pull in the derive feature out of symmetry with the feature above that pulls in the `serde` dependency with `serde_derive`. 

I believe I can remove it though, as it's not strictly required, but then an argument could be made that the above `features =` needs to go as well. 

Since this is not my codebase, I valued symmetry over my own opinions.

AlisCode replied:
@FintanH Do you have any strong opinion on the above ?

AlisCode replied:
Ended up removing the features after removing the newtype in the tests as it wasnt necessary.

FintanH replied:
Sweet, ya I think it makes sense not to have it if it's not needed :+1:

AlisCode replied:
Note: I reverted to using the proc-macro everywhere for easier maintainance, so we actually do need that feature now

greyblake commented on src/lib.rs:
```diff
@@ -1163,4 +1177,37 @@ mod tests {
+        use arbitrary::{Arbitrary, Unstructured};
+
+        #[derive(Debug, Eq, PartialEq, arbitrary::Arbitrary)]
+        pub struct SimpleArbitrary(pub i32);
```
You don't really need this newtype. 
You could use something trivial like `u8` or `i32` (meaning `NonEmpty<i32>` instead of `NonEmpty<SimpleArbitrary>`.

AlisCode replied:
(Same response as above, I do agree though)

FintanH commented on src/lib.rs:
```diff
@@ -1163,4 +1177,37 @@ mod tests {
+        fn test_arbitrary_empty_tail() -> arbitrary::Result<()> {
+            let mut u = Unstructured::new(&[1, 2, 3, 4]);
+            let ne = NonEmpty::<SimpleArbitrary>::arbitrary(&mut u)?;
+            assert!(!ne.is_empty());
+            assert_eq!(ne[0], SimpleArbitrary(67305985));
```
I think it would be clearer for me if the assertions were:
```rust
assert!(ne.tail.is_empty());
assert_eq!(ne.head, SimpleArbitrary(67305985));
```

AlisCode replied:
I ended up doing a full `assert_eq!` check which is more simple IMO

AlisCode commented on src/lib.rs:
```diff
@@ -1163,4 +1177,37 @@ mod tests {
+            let mut u = Unstructured::new(&[1, 2, 3, 4]);
+            let ne = NonEmpty::<SimpleArbitrary>::arbitrary(&mut u)?;
+            assert!(!ne.is_empty());
+            assert_eq!(ne[0], SimpleArbitrary(67305985));
```
Correct :)

The value passed inside `Unstructured::new` is a seed that is used to reproduce an Arbitrary test. Typically, you would use an abstraction over Arbitrary to generate a bunch of random seeds for your test to run in a loop. Popular alternatives include [`arbtest`](https://docs.rs/arbtest/latest/arbtest/) and [`heckcheck`](https://docs.rs/heckcheck/latest/heckcheck/).

AlisCode commented:
Thanks for the fast feedback guys, the PR is updated.

AlisCode commented on src/lib.rs:
```diff
@@ -71,6 +71,8 @@
 //! # Features
 //!
 //! * `serialize`: `serde` support.
```
Adressed in [df932ec](https://github.com/cloudhead/nonempty/pull/54/commits/df932ec8991edc749fc712369b890d58da0d6763)

FintanH reviewed (approved):
Looks good to me :+1:

AlisCode commented:
Sorry for the spam here. Turns out my PR wasn't ready at all!

* I've reverted to actually use the proc-macro, because it has a mechanism to prevent problems with recursion. I also believe it makes maintainance easier
* I've corrected a clippy error (not my code, but I got clippy-sniped because it seems like it's been a while since clippy last ran on this code)
* I modified the CI so that the arbitrary feature gets tested
* I've added the support for `Arbitrary` to `nonzero::NonEmpty`, also using the proc-macro

I think this is good now

FintanH reviewed (approved):
Thanks, I appreciate the clippy cleanup too :pray:

FintanH commented:
@cloudhead could you take a look at this and see if you're happy with the changes too?

AlisCode commented:
@cloudhead  @FintanH, sorry for the ping. What are the next steps to get this merged (and released in the best case) ?

FintanH commented:
> sorry for the ping. What are the next steps to get this merged (and released in the best case) ?

No worries, sorry about the delay. Was deep into some work the past few days :) I can merge it now :+1:

greyblake commented:
@FintanH Thank you!
Would you consider releasing a new version?

AlisCode commented:
Huge thanks for your reactivity and work on the maintainance here guys :)

FintanH commented:
> @FintanH Thank you! Would you consider releasing a new version?

You should be good to go with 0.10 :)
